api/rql/internal/predicate/array.go

```diff
@@ -0,0 +1,124 @@
+package predicate
+
+import (
+	"fmt"
+
+	"github.com/puppetlabs/wash/api/rql"
+	"github.com/puppetlabs/wash/api/rql/internal/errz"
+)
+
+func Array() rql.ValuePredicate {
+	return &array{collectionBase{
+		ctype:            "array",
+		elementPredicate: &arrayElement{p: NPE_ValuePredicate()},
+	}}
+}
+
+type array struct {
+	collectionBase
+}
+
+var _ = rql.ValuePredicate(&array{})
+
+type arrayElement struct {
+	selector interface{}
+	p        rql.ValuePredicate
+}
+
+func (p *arrayElement) Marshal() interface{} {
+	var marshalledSelector interface{}
+	switch t := p.selector.(type) {
+	case stringSelector:
+		marshalledSelector = stringSelectorToStringMap[t]
+	case int:
+		marshalledSelector = float64(t)
+	default:
+		// Should never hit this code-path
+		panic(fmt.Sprintf("Unknown selector %T", p.selector))
+	}
+	return []interface{}{marshalledSelector, p.p.Marshal()}
+}
+
+func (p *arrayElement) Unmarshal(input interface{}) error {
+	array, ok := input.([]interface{})
+	formatErrMsg := "element predicate: must be formatted as [<element_selector>, NPE ValuePredicate]"
```

Comment:
Where would they find a list of valid element selectors?

Reply:
The documentation. I do plan on putting out better error messages